### DISCUSSION TITLE
Filtering, Sorting & Pagination with the Data Layer

### DISCUSSION THREAD

**Discussion by kldavis4** [tinacms/tinacms#2811] [Experimental]
We've released an update to the Data Layer which enables the following collection query features:

- Filtering (using the `filter` parameter)
- Sorting (using the `sort` parameter)
- Pagination (using `after`,`before`,`first`,`last`)

Please give it a try and let us know what you think!

```
yarn add @tinacms/cli@latest
```

For detailed instructions check out the [documentation](https://tina.io/docs/graphql/queries/query-documents/).

**Comment by alborotogarcia** [comment]
Great! Are those parameters available in the ExperimentalGetTinaClient api too? I already enabled the data layer and it looks like it is missing in intellisense/vscode..

**Reply by kldavis4** [reply]
unfortunately, I don't believe these are available in the client yet

**Comment by Yash-Singh1** [comment]
Tried to use this with `datetime` and getting the following error:

```
Error: Unable to fetch, errors: 
	Unable to perform query for Filesystem store
```

**Reply by kldavis4** [reply]
`datetime` should be work.. can you push this to a public repo or add me to the repo so I can troubleshoot?

**Reply by Yash-Singh1** [reply]
Ok LOL, I just realized the reason it wasn’t working was that I was on an older version of Tina. Sorry for the confusion.

**Comment by raguila8** [comment]
Hi, can you sort by nested fields with this new data layer? 
For example, if I have the following in my schema:
```typescript
{
  label: "Meta",
  type: 'object',
  name: 'meta',
  fields: [
    {
      label: "Order",
      name: 'order',
      type: 'number',
      required: true,
    },
  ],
},
```
I want to be able to sort by `meta.order`. I am currently trying to do this, but can't figure it out. Is this possible?
This is related to #2735

**Comment by kldavis4** [comment]
That is not currently supported - once #2735 is addressed, it should be
possible. For the moment the workaround is to move the property up a level
to the collection root.

**Comment by robbyph** [comment]
Is there any way to do something like 

```
postConnection(sort:'date') {
    edges {
      node {
        id
        title
        subtitle
        image
        date
        _sys {
          filename
        }
      }
    }
  }
```

But have it return results in opposite order? Currently this query returns the oldest date at the start of the list, which is cool, but not what i need.

**Reply by kldavis4** [reply]
Yes, if you specify `last` it will reverse the order

**Reply by loleg** [reply]
You mean, something like this? 

```
postConnection(sort: "date", last: 1) { ... }
```

I couldn't figure out how to get sorting to work using GraphQL, so had to go with this ugly hack in the `.tsx` file:

```
  const posts = props.data.postConnection.edges;

  // Hard-coded reversing
  const postsReverse = posts.sort((a,b) => (Date.parse(a.node.date) < Date.parse(b.node.date)));
...
   <Posts data={postsReverse} />
```

Yikes!

**Reply by kldavis4** [reply]
Yes, first/last control whether or not you get ascending or descending results and how many. So `sort: 'date', last: 1` should give you the most recent post.

Maybe you can give some more details regarding the issue you had with it?

**Comment by idiglove** [comment]
Hi I'm trying to sort postConnection with this query:
```
{
  postConnection(sort:"title") {
    edges {
      node {
        id
        title
        published_at
      }
    }
  }
}

```

It gives me an empty edges value. Without sort, I get all the posts.

Here's my schema: 
```
schema: {
    collections: [
      {
        label: "Blog Posts",
        name: "post",
        path: "content/posts",
        format: "md",
        fields: [
          {
            type: "string",
            label: "Title",
            name: "title",
          },
          {
            type: "string",
            label: "Published at",
            name: "published_at",
          },
          {
            type: "rich-text",
            label: "Post Body",
            name: "body",
            isBody: true,
          },
        ],
      },
    ],
  },
  ```


What could I be doing wrong?

**Reply by kldavis4** [reply]
What does the markdown frontmatter look like?

**Reply by idiglove** [reply]
```
---
type_of: > 
 article
id: > 
 268719
title: > 
 Contact Form and Send as Email with AWS SES, Netlify and Gatsby
description: > 
 This tutorial assumes you have Netlify, Gatsby and AWS SES set up. SES stands for Simple Email Servic...
published: > 
 true
published_at: > 
 2020-02-25T11:45:06.039Z
slug: > 
 contact-form-and-send-as-email-with-aws-ses-netlify-and-gatsby-ae5
path: > 
 /idiglove/contact-form-and-send-as-email-with-aws-ses-netlify-and-gatsby-ae5
url: > 
 https://dev.to/idiglove/contact-form-and-send-as-email-with-aws-ses-netlify-and-gatsby-ae5
comments_count: > 
 6
public_reactions_count: > 
 16
page_views_count: > 
 1749
published_timestamp: > 
 2020-02-25T11:45:06Z
positive_reactions_count: > 
 16
cover_image: > 
 null
tag_list: > 
 gatsby,react,aws,netlify
canonical_url: > 
 https://dev.to/idiglove/contact-form-and-send-as-email-with-aws-ses-netlify-and-gatsby-ae5
reading_time_minutes: > 
 2
---
```

**Reply by kldavis4** [reply]
Can you jump on our discord (https://discord.com/invite/zumN63Ybpf) and open a new topic in #help ? Makes it easier to troubleshoot things.

**Reply by idiglove** [reply]
Thanks. Started a thread here:
```
https://discord.com/channels/835168149439643678/1166804525181587496
```
